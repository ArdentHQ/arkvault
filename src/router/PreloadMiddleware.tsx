import { Middleware, MiddlewareParameters, RouteItem } from "@/router/router.types";
import { ContactRoutes } from "@/domains/contact/routing";
import { DashboardRoutes } from "@/domains/dashboard/routing";
import { ExchangeRoutes } from "@/domains/exchange/routing";
import { MessageRoutes } from "@/domains/message/routing";
import { NewsRoutes } from "@/domains/news/routing";
import { SettingRoutes } from "@/domains/setting/routing";
import { TransactionRoutes } from "@/domains/transaction/routing";
import { VoteRoutes } from "@/domains/vote/routing";
import { WalletRoutes } from "@/domains/wallet/routing";
import { ProfileRoutes } from "@/domains/profile/routing";
import { isUnit } from "@/utils/test-helpers";

const getPreloadableRoutes = (path: string): RouteItem[] => {
	if (path === "/") {
		return [...DashboardRoutes, ...ProfileRoutes, ...SettingRoutes, ...ContactRoutes];
	}

	if (path.startsWith("/profiles")) {
		return [
			...ExchangeRoutes,
			...MessageRoutes,
			...NewsRoutes,
			...WalletRoutes,
			...TransactionRoutes,
			...VoteRoutes,
		];
	}

	return [];
};

export class PreloadMiddleware implements Middleware {
	handler({ location }: MiddlewareParameters): boolean {
		/* istanbul ignore next */
		if (isUnit()) {
			return true;
		}

		const routes = getPreloadableRoutes(location.pathname);

		if (routes.length === 0) {
			return true;
		}

		for (const route of routes) {
			route.component["preload"]?.();
		}

		return true;
	}
}
