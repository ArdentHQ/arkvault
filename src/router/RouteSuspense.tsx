import React, { createElement, FunctionComponent, Suspense } from "react";

import { PageSkeleton } from "@/app/components/PageSkeleton";
import { ProfilePageSkeleton } from "@/app/components/PageSkeleton/ProfilePageSkeleton";

interface RouteSuspenseProperties {
	skeleton?: FunctionComponent;
	path: string;
	children?: React.ReactNode;
}

export const RouteSuspense = ({ children, skeleton, path }: RouteSuspenseProperties) => {
	let fallback = <PageSkeleton />;
	if (skeleton) {
		fallback = createElement(skeleton);
	} else if (path.startsWith("/profiles")) {
		fallback = <ProfilePageSkeleton />;
	}

	return <Suspense fallback={fallback}>{children}</Suspense>;
};
