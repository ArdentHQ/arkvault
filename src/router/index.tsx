export * from "./router.routes";
export * from "./router.middlewares";
export * from "./RouterView";
