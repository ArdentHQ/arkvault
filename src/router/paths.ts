export enum ProfilePaths {
	Welcome = "/",
	// Navigation Menu
	Dashboard = "/profiles/:profileId/dashboard",
	Exchange = "/profiles/:profileId/exchange",
	// Contacts
	Contacts = "/profiles/:profileId/contacts",
	// Wallet
	WalletDetails = "/profiles/:profileId/wallets/:walletId",
	CreateWallet = "/profiles/:profileId/wallets/create",
	ImportWallet = "/profiles/:profileId/wallets/import",
	ImportWalletLedger = "/profiles/:profileId/wallets/import/ledger",
	WalletGroupPage = "/profiles/:profileId/network/:networkId",
	// Messages
	SignMessage = "/profiles/:profileId/sign-message",
	SignMessageWallet = "/profiles/:profileId/wallets/:walletId/sign-message",
	VerifyMessage = "/profiles/:profileId/verify-message",
	VerifyMessageWallet = "/profiles/:profileId/wallets/:walletId/verify-message",
	// Transactions
	SendRegistration = "/profiles/:profileId/wallets/:walletId/send-registration/:registrationType",
	SendDelegateResignation = "/profiles/:profileId/wallets/:walletId/send-validator-resignation",
	SendTransferWallet = "/profiles/:profileId/wallets/:walletId/send-transfer",
	SendTransfer = "/profiles/:profileId/send-transfer",
	SendVote = "/profiles/:profileId/send-vote",
	SendVoteWallet = "/profiles/:profileId/wallets/:walletId/send-vote",
	SendMultiSignature = "/profiles/:profileId/wallets/:walletId/send-registration/multiSignature",
	SendDelegateRegistration = "/profiles/:profileId/wallets/:walletId/send-registration/delegateRegistration",
	// Exchange
	ExchangeView = "/profiles/:profileId/exchange/view",
	// Profile
	CreateProfile = "/profiles/create",
	ImportProfile = "/profiles/import",
	// Settings
	Settings = "/profiles/:profileId/settings",
	PasswordSettings = "/profiles/:profileId/settings/password",
	ExportSettings = "/profiles/:profileId/settings/export",
	AppearanceSettings = "/profiles/:profileId/settings/appearance",
	GeneralSettings = "/profiles/:profileId/settings/general",
	ServerManagmentSettings = "/profiles/:profileId/settings/servers",
	NetworkManagmentSettings = "/profiles/:profileId/settings/networks",
	// Votes
	Votes = "/profiles/:profileId/votes",
	VotesWallet = "/profiles/:profileId/wallets/:walletId/votes",
}
