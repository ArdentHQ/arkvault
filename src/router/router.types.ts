import { ComponentType, FC, LazyExoticComponent } from "react";
import { RouteProps } from "react-router";
import { Environment } from "@ardenthq/sdk-profiles";
import { History, Location } from "history";
import { PreloadableComponent } from "@/utils/preload-lazy";

interface RouteItem {
	path: string;
	component: LazyExoticComponent<FC<RouteProps>> | PreloadableComponent<ComponentType<unknown>>;
	skeleton?: FC;
	exact: boolean;
}

interface MiddlewareParameters {
	location: Location;
	env: Environment;
	redirect: (url: string) => void;
	history: History;
}

type LocationState = { from?: string } | undefined;

interface Middleware {
	handler(parameters: MiddlewareParameters): boolean;
}

export type { RouteItem, Middleware, MiddlewareParameters, LocationState };
