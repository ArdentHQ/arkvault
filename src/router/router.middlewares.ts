import { Middleware } from "./router.types";
import { PreloadMiddleware } from "@/router/PreloadMiddleware";
import { UrlValidationMiddleware } from "@/router/UrlValidationMiddleware";
import { WalletMiddleware } from "@/domains/wallet/middleware";

export const middlewares: Middleware[] = [
	new UrlValidationMiddleware(),
	new PreloadMiddleware(),
	new WalletMiddleware(),
];
