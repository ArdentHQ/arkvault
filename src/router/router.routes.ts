import { ContactRoutes } from "@/domains/contact/routing";
import { DashboardRoutes } from "@/domains/dashboard/routing";
import { ExchangeRoutes } from "@/domains/exchange/routing";
import { MessageRoutes } from "@/domains/message/routing";
import { MigrationRoutes } from "@/domains/migration/routing";
import { SettingRoutes } from "@/domains/setting/routing";
import { TransactionRoutes } from "@/domains/transaction/routing";
import { VoteRoutes } from "@/domains/vote/routing";
import { WalletRoutes } from "@/domains/wallet/routing";
import { ProfileRoutes } from "@/domains/profile/routing";
import { RouteItem } from "@/router/router.types";

export const routes: RouteItem[] = [
	...ContactRoutes,
	...DashboardRoutes,
	...ExchangeRoutes,
	...MessageRoutes,
	...MigrationRoutes,
	,
	...SettingRoutes,
	...TransactionRoutes,
	...VoteRoutes,
	...WalletRoutes,
	...ProfileRoutes,
];
