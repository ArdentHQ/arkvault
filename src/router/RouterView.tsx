import React, { createElement, FC, useMemo } from "react";
import { Route, Routes, useLocation } from "react-router-dom";
import { styled } from "twin.macro";

import { useEnvironmentContext } from "@/app/contexts";
import { RouteItem, Middleware } from "@/router/router.types";
import { RouteSuspense } from "@/router/RouteSuspense";
import { PreloadableComponent } from "@/utils/preload-lazy";

interface Properties {
	routes: RouteItem[];
	middlewares?: Middleware[];
}

const Wrapper = styled.div();

export const RouterView: React.VFC<Properties> = ({ routes, middlewares = [] }) => {
	const location = useLocation();

	const { env } = useEnvironmentContext();
	const [_redirectUrl, setRedirectUrl] = React.useState<string | undefined>(undefined);

	// const previousPath = useRef("");

	// useEffect(() => {
	// 	history.listen((route) => {
	// 		// @ts-ignore
	// 		if (!previousPath.current || route.location?.pathname !== previousPath.current) {
	// 			// @ts-ignore
	// 			previousPath.current = route.location?.pathname;
	// 			window.scrollTo(0, 0);
	// 		}
	// 	});
	// }, [history]);

	const canActivate = useMemo(
		() =>
			// @ts-ignore
			middlewares.every((middleware) => middleware.handler({ env, location, redirect: setRedirectUrl })),
		[location, middlewares, env],
	);

	return (
		<Routes>
			{routes.map((route, index) => (
				<Route key={index} path={route.path}>
					<RouteSuspense skeleton={route.skeleton} path={route.path}>
						{canActivate ? (
							<Wrapper data-testid="RouterView__wrapper">
								{createElement(route.component as PreloadableComponent<FC<unknown>>)}
							</Wrapper>
						) : (
							<></>
							// <Redirect to={redirectUrl ?? "/"} />
						)}
					</RouteSuspense>
				</Route>
			))}
		</Routes>
	);
};
