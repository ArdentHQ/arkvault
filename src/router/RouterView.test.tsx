/* eslint-disable react/display-name */
import { createHashHistory } from "history";
import React from "react";
import { MemoryRouter, Router, withRouter } from "react-router-dom";

import { render, screen } from "@/utils/testing-library";

import { Middleware, MiddlewareParameters } from "./router.types";
import { RouterView } from "./RouterView";

describe("RouterView", () => {
	const LocationDisplay = withRouter(({ location }) => (
		// @ts-ignore
		<div data-testid="location-display">{location.location?.pathname || location.pathname}</div>
	));

	it("should render", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<RouterView routes={[{ component: () => <h1>Test</h1>, path: "/" }]} />
			</MemoryRouter>,
		);

		expect(screen.getByTestId("RouterView__wrapper")).toHaveTextContent("Test");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should scroll to top on route change", () => {
		const windowSpy = vi.spyOn(window, "scrollTo");

		const history = createHashHistory();
		history.push("/test");

		render(
			<Router history={history}>
				<RouterView
					routes={[
						{ component: () => <h1>Test 1</h1>, path: "/test" },
						{ component: () => <h1>Test 2</h1>, path: "/test2" },
					]}
				/>
			</Router>,
		);

		history.push("/test2");
		history.replace("/test");

		expect(windowSpy).toHaveBeenCalledTimes(2);
	});

	it("should not scroll to top when route does not change", () => {
		const windowSpy = vi.spyOn(window, "scrollTo");

		const history = createHashHistory();
		history.push("/test");

		render(
			<Router history={history}>
				<RouterView routes={[{ component: () => <h1>Test</h1>, path: "/test" }]} />
			</Router>,
		);

		history.push("/test");
		history.replace("/test");

		expect(windowSpy).toHaveBeenCalledTimes(1);
	});

	it("should block /test router", () => {
		const handler = vi.fn(({ location }: MiddlewareParameters) => location.pathname !== "/test");

		const testMiddleware: Middleware = { handler };
		const history = createHashHistory();
		history.replace("/test");

		render(
			<Router history={history}>
				<RouterView
					routes={[
						{ component: () => <h1>Test</h1>, path: "/test" },
						{ component: () => <h1>Home</h1>, path: "/" },
					]}
					middlewares={[testMiddleware]}
				/>
			</Router>,
		);

		expect(handler).toHaveBeenCalledTimes(2);
	});

	it("should block /test route and redirect to a custom url", () => {
		const handler = vi.fn(({ location, redirect }: MiddlewareParameters) => {
			if (location.pathname === "/test") {
				redirect("/custom");
				return false;
			}
			return true;
		});

		const testMiddleware: Middleware = {
			handler,
		};

		const history = createHashHistory();
		history.push("/test");

		render(
			<Router history={history}>
				<>
					<RouterView
						routes={[
							{ component: () => <h1>Test</h1>, path: "/test" },
							{ component: () => <h1>Custom</h1>, path: "/custom" },
						]}
						middlewares={[testMiddleware]}
					/>
					<LocationDisplay />
				</>
			</Router>,
		);

		expect(handler).toHaveBeenCalledTimes(2);
		expect(screen.getByTestId("location-display")).toHaveTextContent("/custom");
	});
});
