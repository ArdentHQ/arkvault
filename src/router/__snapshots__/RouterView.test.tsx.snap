// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterView should render 1`] = `
<DocumentFragment>
  <div
    class="css-0"
    data-testid="RouterView__wrapper"
  >
    <h1>
      Test
    </h1>
  </div>
</DocumentFragment>
`;
